File: ConcatenatedWords4.cpp
#include<iostream>
#include<vector>
#include<unordered_map>
#include<set>
#include<unordered_set>

using namespace std;

class Solution {
public:
    vector<string> findAllConcatenatedWordsInADict(vector<string>& words) {
    	if(words.size() <= 2) return {};
    	unordered_set<string> dict(words.begin(), words.end());
    	vector<string> result;
    	for(string word : words) {
    		if(word.empty()) continue;
    		if(canConcated(word, dict, 0, 0)) result.push_back(word);
		}
		return result;
    }
    
    bool canConcated(string& word, unordered_set<string>& dict, int pos, int count) {
    	if(pos >= word.size() && count >= 2) return true;
    	for(int i = pos; i < word.size(); i++) {
    		string t = word.substr(pos, i - pos + 1);
    		if(dict.find(t) != dict.end() && canConcated(word, dict, i + 1, count + 1)) return true;
		}
		return false;
	}
};

int main() {
	int n;
	cin>>n;
	
	vector<string> words;
	for(int i = 0; i < n; i++) {
		string word;
		cin>>word;
		words.push_back(word);
	}
	
	Solution *solution = new Solution();
	vector<string> result = solution->findAllConcatenatedWordsInADict(words);
	
	for(auto s : result) cout<<s<<"  ";
	
	return 0;
}

File: ConcatenatedWords4.py
class Solution:
    def findAllConcatenatedWordsInADict(self, words):
        def canConcated(word, pos, dict, count):
            if pos >= len(word) and count >= 2:
                return True
            for i in range(pos, len(word)):
                if word[pos: i + 1] in dict and canConcated(word, i + 1, dict, count + 1):
                    return True
            return False

        if len(words) <= 2:
            return []
        dict = set()
        result = []
        words.sort(key=lambda x: len(x))
        for word in words:
            if not word:
                continue
            if canConcated(word, 0, dict, 0):
                result.append(word)
            dict.add(word)
        return result


if __name__ == '__main__':
    n = int(input())
    words = []
    for i in range(n):
        words.append(input())
    solution = Solution()
    print(solution.findAllConcatenatedWordsInADict(words))
